### PR TITLE
Telemetry data of unique vm for older version of power shell task

### PR DESCRIPTION
**Pull request by rokmicrosoft**

### CODE REVIEW

File: Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1 [hunk 1]
```diff
@@ -92,6 +93,25 @@ function Write-DTLServiceDeprecationMessageIfRequired
     }
 }
 
+function Publish-Azure-Telemetry
+ {
+   param([object] $deploymentResponse, [string] $jobId)
```
Comment (kmkumaran) [inline Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1]: for readability keep each parameter in single line.

File: Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1 [hunk 2]
```diff
@@ -92,6 +93,25 @@ function Write-DTLServiceDeprecationMessageIfRequired
     }
 }
 
+function Publish-Azure-Telemetry
```
Comment (kmkumaran) [inline Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1]: Remove the - between Azure and Telemetry.

File: Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1 [hunk 3]
```diff
@@ -92,6 +93,25 @@ function Write-DTLServiceDeprecationMessageIfRequired
...
+ {
+   param([object] $deploymentResponse, [string] $jobId)
+    if($deploymentResponse){
+       $jsonString = -join("{" , 
+       "`"IsAzureVm`" : `"$($deploymentResponse.IsAzureVm)`"" , 
+       "," , 
+       "`"VmUuidHash`" : `"$($deploymentResponse.VmUuidHash)`"" , 
+       "," , 
+       "`"TelemetryError`" : `"$($deploymentResponse.TelemetryError)`"" ,
```
Comment (kmkumaran) [inline Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1]: TelemetryError can be empty if the execution was successful and viceversa if there was error. Check for the members before you reference. If PS is run in StrictMode referencing can fail.

File: Tasks/PowerShellOnTargetMachines/PowerShellJob.ps1 [hunk 4]
```diff
@@ -53,6 +53,55 @@ param (
...
+        }
+        else
+        {
+            $sessionOption = New-PSSessionOption 
+        }
+        $secpasswd = ConvertTo-SecureString  $credential.Password -AsPlainText -Force
+        $psCredential =  New-Object System.Management.Automation.PSCredential($credential.UserName, $secpasswd)
+        if($httpProtocolOption -eq '-UseHttp')
+        {
+            $session = New-PSSession -Computer $fqdn -Port $port -Credential $psCredential -SessionOption ($sessionOption )
```
Comment (chshrikh) [inline Tasks/PowerShellOnTargetMachines/PowerShellJob.ps1]: better to have a time out for New-PSSession  Ref - New-PSSession

File: Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1 (hunk 1 above)
Comment (rokmicrosoft) [reply Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1]: Done

File: Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1 (hunk 2 above)
Comment (rokmicrosoft) [reply Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1]: Done.

File: Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1 (hunk 3 above)
Comment (rokmicrosoft) [reply Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1]: Done. Now Checking field before using it.

File: Tasks/PowerShellOnTargetMachines/PowerShellJob.ps1 (hunk 4 above)
Comment (rokmicrosoft) [reply Tasks/PowerShellOnTargetMachines/PowerShellJob.ps1]: We are ok with default value for telemetry data OpenTimeout : 00:03:00 CancelTimeout : 00:01:00 IdleTimeout : 00:04:00 . So there is no need to explicitly add it.

File: Tasks/PowerShellOnTargetMachines/PowerShellJob.ps1 [hunk 5]
```diff
@@ -53,6 +53,55 @@ param (
...
+        $VmUuidHash  = Invoke-Command -Session  $session -ScriptBlock {
+            $sha = New-Object System.Security.Cryptography.SHA512CryptoServiceProvider
+            $computerDetails = Get-WmiObject -class Win32_ComputerSystemProduct -namespace root\CIMV2
+            $encoding = [system.Text.Encoding]::ASCII
+            $uuidHash = [System.BitConverter]::ToString( $sha.ComputeHash($encoding.GetBytes($computerDetails.UUID)))
+            $uuidHash = $uuidHash  -replace  "-"  , ""
+            return $uuidHash 
+        }
+        $isAzureVm = Invoke-Command -Session  $session -ScriptBlock {
+            (Get-Process -Name 'WindowsAzureGuestAgent' -ErrorAction Ignore) | Select-Object -First 1 |  ForEach-Object {
```
Comment (chshrikh) [inline Tasks/PowerShellOnTargetMachines/PowerShellJob.ps1]: do we put this in try catch on safer side

Comment (chshrikh) [reply Tasks/PowerShellOnTargetMachines/PowerShellJob.ps1]: no what I mean, do we need to put a try catch in ScriptBlock

File: Tasks/PowerShellOnTargetMachines/Utility.ps1 [hunk 6]
```diff
@@ -182,3 +182,10 @@ function Get-ResourcesProperties
 
     return $resourcesPropertyBag
 }
+
+function IsHosted {
+    $tfsurl = $env:system_teamfoundationcollectionuri.ToLower()
```
Comment (chshrikh) [inline Tasks/PowerShellOnTargetMachines/Utility.ps1]: on safer side add null or empty check on $env:system_teamfoundationcollectionuri

Comment (rokmicrosoft) [reply Tasks/PowerShellOnTargetMachines/Utility.ps1]: Added null check

File: Tasks/PowerShellOnTargetMachines/PowerShellJob.ps1 (hunk 5 above)
Comment (rokmicrosoft) [reply Tasks/PowerShellOnTargetMachines/PowerShellJob.ps1]: If script block will throw exception then it will be catched by outer try catch .

**Pull request merged**